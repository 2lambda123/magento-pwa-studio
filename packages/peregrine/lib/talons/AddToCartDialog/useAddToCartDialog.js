import { useCallback, useEffect, useMemo, useState } from 'react';
import { useMutation, useQuery } from '@apollo/client';

import mergeOperations from '../../util/shallowMerge';
import { useCartContext } from '../../context/cart';
import defaultOperations from './addToCartDialog.gql';
import { useEventingContext } from '../../context/eventing';
import { isProductConfigurable } from '@magento/peregrine/lib/util/isProductConfigurable';
import { findAllMatchingVariants } from '@magento/peregrine/lib/util/findAllMatchingVariants';
import { getOutOfStockIndexes } from '@magento/peregrine/lib/util/getOutOfStockIndexes';
import { createProductVariants } from '@magento/peregrine/lib/util/createProductVariants';

const OUT_OF_STOCK_CODE = 'OUT_OF_STOCK';

const getOutOfStockVariants = (
    product,
    optionCodes,
    singleOptionSelection,
    multipleOptionSelections,
    isOutOfStockProductDisplayed
) => {
    const isConfigurable = product && isProductConfigurable(product);
    const singeOptionSelected =
        singleOptionSelection && singleOptionSelection.size === 1;
    const optionsSelected =
        Array.from(multipleOptionSelections.values()).filter(value => !!value)
            .length > 1;
    const outOfStockIndexes = [];

    // Find the combination of k elements in the array.
    // For example: array is [1,2,3]. k=2.
    // The results are [[1,2],[1,3],[2,3]]
    function getCombinations(array, k, prefix = []) {
        if (k == 0) return [prefix];
        return array.flatMap((value, index) =>
            getCombinations(array.slice(index + 1), k - 1, [...prefix, value])
        );
    }

    if (isConfigurable) {
        let variants = product.variants;
        let variantsIfOutOfStockProductsNotDisplayed = createProductVariants(
            product
        );
        //If out of stock products is set to not displayed, use the variants created
        variants = isOutOfStockProductDisplayed
            ? variants
            : variantsIfOutOfStockProductsNotDisplayed;

        const numberOfVariations = variants[0].attributes.length;
        const selectedIndexes = Array.from(
            multipleOptionSelections.values()
        ).flat();

        // If only one pair of variations, display out of stock variations before option selection
        if (numberOfVariations === 1) {
            let outOfStockOptions = variants.filter(
                variant => variant.product.stock_status === OUT_OF_STOCK_CODE
            );

            let outOfStockIndex = outOfStockOptions.map(option =>
                option.attributes.map(attribute => attribute.value_index)
            );
            return outOfStockIndex;
        } else {
            if (singeOptionSelected) {
                const items = findAllMatchingVariants({
                    optionCodes,
                    singleOptionSelection,
                    variants
                });
                const outOfStockItemsIndexes = getOutOfStockIndexes(items);

                // For all the out of stock options associated with current selection, display out of stock swatches
                // when the number of matching indexes of selected indexes and out of stock indexes are not smaller than the total groups of swatches minus 1
                for (const indexes of outOfStockItemsIndexes) {
                    const sameIndexes = indexes.filter(num =>
                        selectedIndexes.includes(num)
                    );
                    const differentIndexes = indexes.filter(
                        num => !selectedIndexes.includes(num)
                    );
                    if (sameIndexes.length >= optionCodes.size - 1) {
                        outOfStockIndexes.push(differentIndexes);
                    }
                }
                // Display all possible out of stock swatches with current selections, when all groups of swatches are selected
                if (
                    optionsSelected &&
                    selectedIndexes.length === optionCodes.size
                ) {
                    const selectedIndexesCombinations = getCombinations(
                        selectedIndexes,
                        selectedIndexes.length - 1
                    );
                    // Find out of stock items and indexes for each combination
                    let oosIndexes = [];
                    for (const option of selectedIndexesCombinations) {
                        // Map the option indexes to their optionCodes
                        const curOption = new Map(
                            [...multipleOptionSelections].filter(([key, val]) =>
                                option.includes(val)
                            )
                        );
                        const curItems = findAllMatchingVariants({
                            optionCodes: optionCodes,
                            singleOptionSelection: curOption,
                            variants: variants
                        });
                        const outOfStockIndex = getOutOfStockIndexes(curItems)
                            .flat()
                            .filter(idx => !selectedIndexes.includes(idx));
                        oosIndexes.push(outOfStockIndex);
                    }
                    return oosIndexes;
                }
                return outOfStockIndexes;
            }
        }
    }
    return [];
};

export const useAddToCartDialog = props => {
    const { item, onClose } = props;
    const sku = item && item.product.sku;

    const [, { dispatch }] = useEventingContext();

    const operations = mergeOperations(defaultOperations, props.operations);

    const [userSelectedOptions, setUserSelectedOptions] = useState(new Map());
    const [currentImage, setCurrentImage] = useState();
    const [currentPrice, setCurrentPrice] = useState();
    const [currentDiscount, setCurrentDiscount] = useState();
    const [singleOptionSelection, setSingleOptionSelection] = useState();
    const [multipleOptionSelections, setMultipleOptionSelections] = useState(
        new Map()
    );

    const [{ cartId }] = useCartContext();

    const optionCodes = useMemo(() => {
        const optionCodeMap = new Map();
        if (item) {
            item.product.configurable_options.forEach(option => {
                optionCodeMap.set(option.attribute_id, option.attribute_code);
            });
        }
        return optionCodeMap;
    }, [item]);

    // Check if display out of stock products option is selected in the Admin Dashboard
    const isOutOfStockProductDisplayed = useMemo(() => {
        if (item) {
            let totalVariants = 1;
            const { product } = item;
            const isConfigurable = isProductConfigurable(product);
            if (product.configurable_options && isConfigurable) {
                for (const option of product.configurable_options) {
                    const length = option.values.length;
                    totalVariants = totalVariants * length;
                }
                return product.variants.length === totalVariants;
            }
        }
    }, [item]);

    const outOfStockVariants = useMemo(() => {
        if (item) {
            const product = item.product;
            return getOutOfStockVariants(
                product,
                optionCodes,
                singleOptionSelection,
                multipleOptionSelections,
                isOutOfStockProductDisplayed
            );
        }
    }, [
        item,
        optionCodes,
        singleOptionSelection,
        multipleOptionSelections,
        isOutOfStockProductDisplayed
    ]);

    const selectedOptionsArray = useMemo(() => {
        if (item) {
            const existingOptionsMap = item.configurable_options.reduce(
                (optionsMap, selectedOption) => {
                    return optionsMap.set(
                        selectedOption.id,
                        selectedOption.value_id
                    );
                },
                new Map()
            );
            const mergedOptionsMap = new Map([
                ...existingOptionsMap,
                ...userSelectedOptions
            ]);

            const selectedOptions = [];
            mergedOptionsMap.forEach((selectedValueId, attributeId) => {
                const configurableOption = item.product.configurable_options.find(
                    option => option.attribute_id_v2 === attributeId
                );
                const configurableOptionValue = configurableOption.values.find(
                    optionValue => optionValue.value_index === selectedValueId
                );

                selectedOptions.push(configurableOptionValue.uid);
            });

            return selectedOptions;
        }

        return [];
    }, [item, userSelectedOptions]);

    const { data, loading: isFetchingProductDetail } = useQuery(
        operations.getProductDetailQuery,
        {
            fetchPolicy: 'cache-and-network',
            nextFetchPolicy: 'cache-first',
            variables: {
                configurableOptionValues: selectedOptionsArray,
                sku
            },
            skip: !sku
        }
    );

    const [
        addProductToCart,
        { error: addProductToCartError, loading: isAddingToCart }
    ] = useMutation(operations.addProductToCartMutation);

    useEffect(() => {
        if (data) {
            const product = data.products.items[0];
            const {
                media_gallery: selectedProductMediaGallery,
                variant: selectedVariant
            } = product.configurable_product_options_selection;

            const currentImage =
                selectedProductMediaGallery.length &&
                selectedOptionsArray.length
                    ? selectedProductMediaGallery[0]
                    : product.image;

            setCurrentImage(currentImage);

            const finalPrice = selectedVariant
                ? selectedVariant.price_range.maximum_price.final_price
                : product.price_range.maximum_price.final_price;

            const discount = selectedVariant
                ? selectedVariant.price_range.maximum_price.discount
                : product.price_range.maximum_price.discount;

            setCurrentDiscount(discount);
            setCurrentPrice(finalPrice);
        }
    }, [data, selectedOptionsArray.length]);

    const handleOnClose = useCallback(() => {
        onClose();
        setCurrentImage();
        setCurrentPrice();
        setUserSelectedOptions(new Map());
        setMultipleOptionSelections(new Map());
    }, [onClose]);

    const handleOptionSelection = useCallback(
        (optionId, value) => {
            setUserSelectedOptions(existing =>
                new Map(existing).set(parseInt(optionId), value)
            );
            // Create a new Map to keep track of user single selection with key as String
            const nextSingleOptionSelection = new Map();
            nextSingleOptionSelection.set(optionId, value);
            setSingleOptionSelection(nextSingleOptionSelection);
            // Create a new Map to keep track of multiple selections with key as String
            const nextMultipleOptionSelections = new Map([
                ...multipleOptionSelections
            ]);
            nextMultipleOptionSelections.set(optionId, value);
            setMultipleOptionSelections(nextMultipleOptionSelections);
        },
        [singleOptionSelection, multipleOptionSelections]
    );

    const handleAddToCart = useCallback(async () => {
        try {
            const quantity = 1;

            await addProductToCart({
                variables: {
                    cartId,
                    cartItem: {
                        quantity,
                        selected_options: selectedOptionsArray,
                        sku
                    }
                }
            });

            const selectedOptionsLabels =
                selectedOptionsArray?.map((value, i) => ({
                    attribute: item.product.configurable_options[i].label,
                    value:
                        item.product.configurable_options[i].values.find(
                            x => x.uid === value
                        )?.label || null
                })) || null;

            dispatch({
                type: 'CART_ADD_ITEM',
                payload: {
                    cartId,
                    sku: item.product.sku,
                    name: item.product.name,
                    priceTotal: currentPrice.value,
                    currencyCode: currentPrice.currency,
                    discountAmount: currentDiscount.amount_off,
                    selectedOptions: selectedOptionsLabels,
                    quantity
                }
            });

            handleOnClose();
        } catch (error) {
            console.error(error);
        }
    }, [
        addProductToCart,
        cartId,
        currentDiscount,
        currentPrice,
        dispatch,
        handleOnClose,
        item,
        selectedOptionsArray,
        sku
    ]);

    const imageProps = useMemo(() => {
        if (currentImage) {
            return {
                alt: currentImage.label,
                src: currentImage.url,
                width: 400
            };
        }
    }, [currentImage]);

    const priceProps = useMemo(() => {
        if (currentPrice) {
            return {
                currencyCode: currentPrice.currency,
                value: currentPrice.value
            };
        }
    }, [currentPrice]);

    const configurableOptionProps = useMemo(() => {
        if (item) {
            return {
                onSelectionChange: handleOptionSelection,
                options: item.product.configurable_options,
                selectedValues: item.configurable_options
            };
        }
    }, [handleOptionSelection, item]);

    const buttonProps = useMemo(() => {
        if (item) {
            return {
                disabled:
                    item.product.configurable_options.length !==
                        selectedOptionsArray.length || isAddingToCart,
                onClick: handleAddToCart,
                priority: 'high'
            };
        }
    }, [handleAddToCart, isAddingToCart, item, selectedOptionsArray.length]);

    return {
        buttonProps,
        configurableOptionProps,
        formErrors: [addProductToCartError],
        handleOnClose,
        outOfStockVariants,
        imageProps,
        isFetchingProductDetail,
        priceProps
    };
};
