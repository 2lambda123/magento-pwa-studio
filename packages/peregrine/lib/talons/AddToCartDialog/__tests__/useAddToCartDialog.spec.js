import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { act, renderHook } from '@testing-library/react-hooks';

import defaultOperations from '../addToCartDialog.gql';
import { useAddToCartDialog } from '../useAddToCartDialog';

jest.mock('../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

const productMock = {
    __typename: 'ConfigurableProduct',
    configurable_options: [],
    configurable_product_options_selection: {
        media_gallery: [
            {
                label: 'red-shirt',
                url: 'https://example.com/media/red-shirt.jpg'
            }
        ],
        variant: null
    },
    id: 1149,
    image: {
        label: 'default-shirt',
        url: 'https://example.com/media/default-shirt.jpg'
    },
    price_range: {
        maximum_price: {
            final_price: {
                currency: 'USD',
                value: 123.45
            }
        }
    },
    uid: 'UID001=='
};

const getProductDetailMock1 = {
    request: {
        query: defaultOperations.getProductDetailQuery,
        variables: {
            configurableOptionValues: ['red-uid'],
            sku: 'nice-shirt'
        }
    },
    result: {
        data: {
            products: {
                items: [productMock]
            }
        }
    }
};

const getProductDetailMock2 = {
    request: {
        query: defaultOperations.getProductDetailQuery,
        variables: {
            configurableOptionValues: ['red-uid', 'medium-uid'],
            sku: 'nice-shirt'
        }
    },
    result: {
        data: {
            products: {
                items: [
                    {
                        ...productMock,
                        configurable_product_options_selection: {
                            media_gallery: [],
                            variant: {
                                id: 925,
                                price_range: {
                                    maximum_price: {
                                        final_price: {
                                            currency: 'EUR',
                                            value: 456.78
                                        }
                                    }
                                },
                                uid: 'OTI1'
                            }
                        }
                    }
                ]
            }
        }
    }
};

const mockItem = {
    configurable_options: [{ id: 1, value_id: 'red-id' }],
    product: {
        configurable_options: [
            {
                attribute_id_v2: 1,
                values: [
                    { value_index: 'red-id', uid: 'red-uid' },
                    { value_index: 'blue-id', uid: 'blue-uid' }
                ]
            },
            {
                attribute_id_v2: 2,
                values: [
                    { value_index: 'large-id', uid: 'large-uid' },
                    { value_index: 'medium-id', uid: 'medium-uid' }
                ]
            }
        ],
        sku: 'nice-shirt'
    }
};

const initialProps = {
    onClose: jest.fn().mockName('props.onClose')
};

const renderHookWithProviders = ({
    renderHookOptions = { initialProps },
    mocks = [getProductDetailMock1, getProductDetailMock2]
} = {}) => {
    const wrapper = ({ children }) => (
        <MockedProvider mocks={mocks} addTypename={false}>
            {children}
        </MockedProvider>
    );

    return renderHook(useAddToCartDialog, { wrapper, ...renderHookOptions });
};

test('returns correct shape when closed', () => {
    const { result } = renderHookWithProviders();

    expect(result.current).toMatchInlineSnapshot(`
        Object {
          "buttonProps": undefined,
          "configurableOptionProps": undefined,
          "formErrors": Array [
            undefined,
          ],
          "handleOnClose": [Function],
          "imageProps": undefined,
          "isFetchingProductDetail": false,
          "priceProps": undefined,
        }
    `);
});

test('returns configurable option props', async () => {
    const { result } = renderHookWithProviders({
        renderHookOptions: { initialProps: { item: mockItem } }
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(result.current.configurableOptionProps).toMatchSnapshot();
});

test('returns price props', async () => {
    const { result } = renderHookWithProviders({
        renderHookOptions: { initialProps: { item: mockItem } }
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(result.current.priceProps).toMatchInlineSnapshot(`
        Object {
          "currencyCode": "USD",
          "value": 123.45,
        }
    `);

    act(() => {
        result.current.configurableOptionProps.onSelectionChange(
            '2',
            'medium-id'
        );
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(result.current.priceProps).toMatchInlineSnapshot(`
        Object {
          "currencyCode": "EUR",
          "value": 456.78,
        }
    `);
});

test('returns image props', async () => {
    const { result } = renderHookWithProviders({
        renderHookOptions: { initialProps: { item: mockItem } }
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(result.current.imageProps).toMatchInlineSnapshot(`
        Object {
          "alt": "red-shirt",
          "src": "https://example.com/media/red-shirt.jpg",
          "width": 400,
        }
    `);

    act(() => {
        result.current.configurableOptionProps.onSelectionChange(
            '2',
            'medium-id'
        );
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(result.current.imageProps).toMatchInlineSnapshot(`
        Object {
          "alt": "default-shirt",
          "src": "https://example.com/media/default-shirt.jpg",
          "width": 400,
        }
    `);
});

test('closing dialog resets state', async () => {
    const { result } = renderHookWithProviders({
        renderHookOptions: { initialProps: { ...initialProps, item: mockItem } }
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(initialProps.onClose).not.toHaveBeenCalled();
    expect(result.current.imageProps).toBeTruthy();
    expect(result.current.priceProps).toBeTruthy();

    act(() => {
        result.current.handleOnClose();
    });

    expect(initialProps.onClose).toHaveBeenCalled();
    expect(result.current.imageProps).toBeUndefined();
    expect(result.current.priceProps).toBeUndefined();
});

test('addToCart succeeds and closes dialog', async () => {
    const addToCartMock = {
        request: {
            query: defaultOperations.addProductToCartMutation,
            variables: {
                cartId: '123',
                cartItem: {
                    quantity: 1,
                    selected_options: ['red-uid', 'medium-uid'],
                    sku: 'nice-shirt'
                }
            }
        },
        result: {
            data: {}
        }
    };

    const { result } = renderHookWithProviders({
        renderHookOptions: {
            initialProps: { ...initialProps, item: mockItem }
        },
        mocks: [
            getProductDetailMock1,
            getProductDetailMock1,
            getProductDetailMock2,
            addToCartMock
        ]
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    act(() => {
        result.current.configurableOptionProps.onSelectionChange(
            '2',
            'medium-id'
        );
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(initialProps.onClose).not.toHaveBeenCalled();
    expect(result.current.imageProps).toBeTruthy();
    expect(result.current.priceProps).toBeTruthy();

    act(() => {
        result.current.buttonProps.onClick();
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(initialProps.onClose).toHaveBeenCalled();
});

test('addToCart failures returns error', async () => {
    const addToCartMock = {
        request: {
            query: defaultOperations.addProductToCartMutation,
            variables: {
                cartId: '123',
                cartItem: {
                    quantity: 1,
                    selected_options: ['red-uid', 'medium-uid'],
                    sku: 'nice-shirt'
                }
            }
        },
        error: new Error('Oh noes! Something went wrong :(')
    };

    const { result } = renderHookWithProviders({
        renderHookOptions: {
            initialProps: { ...initialProps, item: mockItem }
        },
        mocks: [getProductDetailMock1, getProductDetailMock2, addToCartMock]
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    act(() => {
        result.current.configurableOptionProps.onSelectionChange(
            '2',
            'medium-id'
        );
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    act(() => {
        result.current.buttonProps.onClick();
    });

    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
    });

    expect(initialProps.onClose).not.toHaveBeenCalled();
    expect(result.current.formErrors).toMatchInlineSnapshot(`
        Array [
          [Error: Oh noes! Something went wrong :(],
        ]
    `);
});
