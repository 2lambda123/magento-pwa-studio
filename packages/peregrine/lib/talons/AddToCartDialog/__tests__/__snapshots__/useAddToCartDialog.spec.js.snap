// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns configurable option props 1`] = `
Object {
  "onSelectionChange": [Function],
  "options": Array [
    Object {
      "attribute_id_v2": 1,
      "values": Array [
        Object {
          "uid": "red-uid",
          "value_index": "red-id",
        },
        Object {
          "uid": "blue-uid",
          "value_index": "blue-id",
        },
      ],
    },
    Object {
      "attribute_id_v2": 2,
      "values": Array [
        Object {
          "uid": "large-uid",
          "value_index": "large-id",
        },
        Object {
          "uid": "medium-uid",
          "value_index": "medium-id",
        },
      ],
    },
  ],
  "selectedValues": Array [
    Object {
      "id": 1,
      "value_id": "red-id",
    },
  ],
}
`;
